hack/lib/start.sh

```diff
@@ -0,0 +1,587 @@
+	echo "[INFO] Starting etcd"
+	local openshift_env=( "OPENSHIFT_ON_PANIC=crash" )
+	local openshift_executable
+	openshift_executable="$(os::start::internal::openshift_executable)"
+	${openshift_executable} start etcd \
+		--config="${MASTER_CONFIG_DIR}/master-config.yaml" &>"${LOG_DIR}/etcd.log" &
+	export ETCD_PID=$!
+
+	echo "[INFO] etcd server start at: "
+	date 
+
+	wait_for_url "${API_SCHEME}://${API_HOST}:${ETCD_PORT}/version" "etcd: " 0.25 80
+
+	echo "[INFO] etcd server health checks done at: "
+	date
+}
+readonly -f os::start::etcd
+
+# os::start::api_server starts the OpenShift API server
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - ARTIFACT_DIR
+#  - MASTER_CONFIG_DIR
+#  - API_SCHEME
+#  - API_HOST
+#  - API_PORT
+#  - KUBELET_SCHEME
+#  - KUBELET_HOST
+#  - KUBELET_PORT
+# Arguments:
+# 1 - api server version
+# Returns:
+#  - export OS_PID
+function os::start::api_server() {
+	local api_server_version=${1:-}
+	local openshift_volumes=( "${MASTER_CONFIG_DIR}" )
+	local openshift_env=( "OPENSHIFT_PROFILE=web" "OPENSHIFT_ON_PANIC=crash" )
+	local openshift_executable
+	openshift_executable="$(os::start::internal::openshift_executable "${api_server_version}")"
+
+	${openshift_executable} start master api \
+		--config="${MASTER_CONFIG_DIR}/master-config.yaml" \
+	&>"${LOG_DIR}/apiserver.log" &
+
+	export API_SERVER_PID=$!
+
+	echo "[INFO] OpenShift API server start at: "
+	date
+
+	wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/healthz" "apiserver: " 0.25 80
+	wait_for_url "${API_SCHEME}://${API_HOST}:${API_PORT}/healthz/ready" "apiserver(ready): " 0.25 160
+
+	echo "[INFO] OpenShift API server health checks done at: "
+	date
+}
+readonly -f os::start::api_server
+
+# os::start::controllers starts the OpenShift controllers
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - MASTER_CONFIG_DIR
+# Arguments:
+# 1 - controllers version
+# Returns:
+#  - export CONTROLLERS_PID
+function os::start::controllers() {
+	local controllers_version=${1:-}
+	local openshift_volumes=( "${MASTER_CONFIG_DIR}" )
+	local openshift_env=( "OPENSHIFT_ON_PANIC=crash" )
+	local openshift_executable
+	openshift_executable="$(os::start::internal::openshift_executable "${controllers_version}")"
+
+	${openshift_executable} start master controllers \
+		--config="${MASTER_CONFIG_DIR}/master-config.yaml" \
+	&>"${LOG_DIR}/controllers.log" &
+
+	export CONTROLLERS_PID=$!
+
+	echo "[INFO] OpenShift controllers start at: "
+	date
+}
+readonly -f os::start::controllers
+
+
+# os::start::internal::start_node starts the OpenShift node
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - USE_LATEST_IMAGES
+#  - NODE_CONFIG_DIR
+#  - KUBELET_SCHEME
+#  - KUBELET_HOST
+#  - KUBELET_PORT
+# Arguments:
+#    none
+# Returns:
+#  - export NODE_PID
+function os::start::internal::start_node() {
+	local use_latest_images
+	if [[ -n "${USE_LATEST_IMAGES:-}" ]]; then
+		use_latest_images="true"
+	else
+		use_latest_images="false"
+	fi
+
+	mkdir -p ${LOG_DIR}
```

Comment:
I'm in favor of it being out of scope :)
And calling mkdir -p has a zero net effect if the dir already exists, so I say there's no harm